### PR TITLE
load config from toml + other changes

### PR DESCRIPTION
**Pull request by onyx-and-iris**
Hi.

I've made a few changes to the code. Namely:

- rename OBS::WebSocket to OBS::WS to avoid namespace conflict with websocket/driver ? Perhaps there is a better way to do this?
- Add support for loading configuration from a config.toml file, but still allow passing cmdline arguments.
- Separate the code in module OBS::WS::Protocol into individual files.
- Added OBS::Connection module that holds SocketWrapper class and connect function.
- start_driver method added to Client class.
- run method in Client class to automatically start the driver if a block is passed.
- an example connecting ruby-obs-websocket with voicemeeter
- all files run through prettier

The purpose of these changes is to simplify the code from end-user perspective.

If any of these changes fit with your ideas for the code feel free to include them. If you decide not to merge I would ask kindly if you would include a way to load connection info from a configuration file, toml or otherwise.

Thank you for your time.

### CODE REVIEW

**Comment by hanazuki** [comment]
Thank you for being interested in contributing to this project. But your patch is quite large and contains too many different kinds of changes in one commit, which takes time for me to review. So the patch is not acceptable in its current form, but I think some of your ideas are very interesting and worth accepting.

If you are willing to revise your patch, please organize your changes into multiple commits so that each commit introduces a single feature, and submit them as separate Pull Requests for easier review. Thank you!

My thoughts on your ideas (not on your patch):

> rename OBS::WebSocket to OBS::WS to avoid namespace conflict with websocket/driver ? Perhaps there is a better way to do this?

This should not be necessary. When both of the constants `OBS::WebSocket` from this gem and `WebSocket` from `websocket/driver` are in scope, you can distinguish them by using `OBS::WebSocket` and `::WebSocket`, respectively.

> Add support for loading configuration from a config.toml file, but still allow passing cmdline arguments.

I think this is a good enhancement for the library consumer. The configuration format is however up to the consumer; you prefer TOML, others YAML, and yet another XML. This library should not assume any one of them or depend on them.

An acceptable alternative would be to have a method that takes keyword arguments (or a hash) and set up a connection such as 
```ruby
def connect(uri:, password: nil)
  # setup a TCP socket and construct Client instance
end
```
With this method, users could use their preferred way to parse their configuration file:
```ruby
connect(**JSON.load_file('config.json', symbolize_names: true))
# or
connect(**PerfectTOML.load_file('config.toml', symbolize_names: true))
```

> Separate the code in module OBS::WS::Protocol into individual files.

Because the file `protocol.rb` is programmatically generated from the upstream API definition, please don't edit it (or split it) manually. Instead, you can make changes in `codegen/generate-protocol` to do so.

But I don't think there is a reason to split the file, as it should never be manually edited. For reference purposes, you can run `rake doc` to generate HTML documentation.

> Added OBS::Connection module that holds SocketWrapper class and connect function.
> start_driver method added to Client class.
> run method in Client class to automatically start the driver if a block is passed.

+1 for the idea.

When I first designed this library, I wanted it not to depend on any specific WebSocket client implementation because there was no best one. So I intentionally omitted the connection-management feature from this library and delegated it to the library consumer.

But now I think `websocket/drive` is enough simple and suits common needs. And it is not easy to integrate other I/O schedulers such as EventMachine with Concurrent Ruby, on which this library heavily depends. So I'm for to add runtime dependency on `websocket/driver` and to introduce the connection-management feature specific for `websocket/driver`.

For the implementation: 
I prefer `OBS::WebSocket::Connection` to be a Class but not a Module. It should not have any global state and be able to connect to multiple instances of OBS Studios simultaneously.
Also `Client` class should not touch WebSocket::Driver or TCPSocket. It makes us easier to maintain the code if connection management is contained in the `Connection` class.

> an example connecting ruby-obs-websocket with voicemeeter

A real-world example is a nice-to-have.

> all files run through prettier

Please don't. Massive code formatting makes it impossible to review the change and makes it very hard to follow the code history. Furthermore, without proper editor support in the maintainer's environment, the format is soon rolled back to its original state.

**Comment by onyx-and-iris** [comment]
Hello. Thank you for your detailed response.

I figured there was a way to avoid the namespace conflict I just wasn't sure the correct way.

Regarding configuration, yes that is a good suggestion, a connect method that accepts keyword arguments. Let the user decide on their preferred format. That would satisfy my request without restricting the format type. 

I will take a look into codegen/generate-protocol and rake doc.

Thank you for explaining why the connection management features are delegated to the user. Yes, if that can be integrated somehow that would be best imo.

Regarding prettier, understood. It was a nuisance to setup on vscode.

If I can figure out a way to implement my ideas in an acceptable way I will make some pull requests.

Again, thank you for your time.

I will close this PR now.
